cc/keystone/middleware/lifesaver.py

```diff
@@ -145,12 +149,79 @@ def get_user(self, request):
 
         return {'user': self.utils.normalize(user), 'domain': self.utils.normalize(domain)}
 
+
+    def get_token(self, request: Request) -> None | str:
+        """
+        Tries to fetch token id from the request
+        :param request: the clients request
+        :return: str 'token_id'
+        """
+
+        # shortcut for version discovery request
+        if '/v3/' == request.path:
+            return None
+
+        token_id = None
+        try:
+            if '/v3/auth/tokens' == request.path and 'POST' == request.method:
+                body = request.json_body
+                auth = body.get('auth', {})
+                identity = auth.get('identity', {})
+                token = identity.get('token', {})
+                token_id = token.get('id', None)
+            elif '/v3/auth/tokens' == request.path and 'GET' == request.method:
+                if "X-Subject-Token" in request.headers.keys():
+                    token_id = request.headers.get("X-Subject-Token")
```

Comment:
A token could be
* valid
* invalid (e.g. never issued, or empty)
* expired
* revoked

And such tokens could appear in
* in an request authentication payload
  * `X-Auth-Token` header
  * or in body
* and/or in `X-Subject-Token` header

It's not clear if the code handles all reasonable combinations of those tokens - does it?